File: SiderShopping/app/src/main/java/com/mmart/sidershopping/view/productlist/ProductListAdapter.java
package com.mmart.sidershopping.view.productlist;

import android.arch.paging.PagedListAdapter;
import android.content.Context;
import android.support.annotation.NonNull;
import android.support.v7.util.DiffUtil;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.TextView;

import com.mmart.sidershopping.ItemClickListener;
import com.mmart.sidershopping.R;
import com.mmart.sidershopping.model.entity.Product;
import com.squareup.picasso.Picasso;

import java.text.DecimalFormat;

public class ProductListAdapter extends PagedListAdapter<Product, ProductListAdapter.ProductViewHolder> {

    private Context context;
    private static DecimalFormat df2 = new DecimalFormat("#.##");
    private static ItemClickListener itemClickListener;

    protected ProductListAdapter(Context context) {
        super(DIFF_CALLBACK);
        this.context = context;
    }

    @NonNull
    @Override
    public ProductViewHolder onCreateViewHolder(@NonNull ViewGroup viewGroup, int i) {
        View view = LayoutInflater.from(context).inflate(R.layout.product_item_list, viewGroup, false);

        ProductViewHolder viewHolder = new ProductListAdapter.ProductViewHolder(view);
        return viewHolder;
    }

    @Override
    public void onBindViewHolder(@NonNull ProductViewHolder holder, int position) {
        Product movie = getItem(position);

        if (movie != null) {

            holder.bindTo(movie);
        }
    }

    private static DiffUtil.ItemCallback<Product> DIFF_CALLBACK =
            new DiffUtil.ItemCallback<Product>() {
                @Override
                public boolean areItemsTheSame(@NonNull Product product, @NonNull Product t1) {
                    return product.getId() == t1.getId();
                }

                @Override
                public boolean areContentsTheSame(@NonNull Product product, @NonNull Product t1) {
                    return product.equals(t1);
                }

            };

    public class ProductViewHolder extends RecyclerView.ViewHolder implements View.OnClickListener  {

        ImageView image;
        TextView name;
        TextView oldPrice;
        TextView newPrice;
        Product currentProduct;

        public ProductViewHolder(@NonNull View itemView) {
            super(itemView);

            image = itemView.findViewById(R.id.product_image);
            name = itemView.findViewById(R.id.product_name);
            newPrice = itemView.findViewById(R.id.product_new_price);
            oldPrice = itemView.findViewById(R.id.product_old_price);

            itemView.setOnClickListener(this);

        }

        public void bindTo(Product product) {
            currentProduct = product;

            Picasso.get()
                    .load(product.getUrlImagem())
                    .fit()
                    .centerCrop()
                    .into(image);

            name.setText(product.getNome());
            oldPrice.setText("De: " + df2.format(product.getPrecoDe()));
            newPrice.setText("Por " + df2.format(product.getPrecoPor()));
        }

        @Override
        public void onClick(View view) {
            if(itemClickListener != null) {
                itemClickListener.onItemClick(currentProduct);
            }
        }
    }

    public void setOnItemClickListener(ItemClickListener itemClickListener){
        this.itemClickListener = itemClickListener;
    }
}

File: app/src/main/java/com/eric/alodjinha/features/product/adapter/ProductListAdapter.kt
package com.eric.alodjinha.features.product.adapter

import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.recyclerview.widget.RecyclerView
import com.eric.alodjinha.R
import com.eric.alodjinha.features.product.model.Product
import com.squareup.picasso.Picasso
import kotlinx.android.synthetic.main.item_products.view.*

class ProductListAdapter(var categories: List<Product>) : RecyclerView.Adapter<ProductListAdapter.ViewHolder>() {

    var onClick: (Product) -> Unit = {}

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ViewHolder {
        return ViewHolder(
            LayoutInflater.from(parent.context)
                .inflate(R.layout.item_products, parent, false)
        )
    }

    override fun onBindViewHolder(holder: ViewHolder, position: Int) {
        val product = categories[position]

        Picasso.get()
            .load(product.urlImagem)
            .placeholder(R.drawable.shopping_bag)
            .error(R.drawable.warning)
            .into(holder.imageViewProduct)

        holder.textViewProductName.text = product.nome
        holder.textViewPriceFrom.text = product.precoDe.toString()
        holder.textViewPriceBy.text = product.precoPor.toString()
        holder.itemView.setOnClickListener { onClick(product) }
    }

    override fun getItemCount(): Int {
        return categories.size
    }

    class ViewHolder(val view: View) : RecyclerView.ViewHolder(view) {
        val imageViewProduct = view.imageViewProduct
        val textViewProductName = view.textViewProductName
        val textViewPriceFrom = view.textViewPriceFrom
        val textViewPriceBy = view.textViewPriceBy
    }
}